ipywidgets/static/widgets/js/manager.js

```diff
@@ -418,30 +430,34 @@ define([
         return this._get_connected_kernel().then(function(kernel) {
 
             // Recreate all the widget models for the given notebook state.
-            var all_models = Promise.all(_.map(Object.keys(state), function (model_id) {
-                // Recreate a comm using the widget's model id (model_id == comm_id).
-                var new_comm = new comm.Comm(kernel.widget_manager.comm_target_name, model_id);
-                kernel.comm_manager.register_comm(new_comm);
-
-                // Create the model using the recreated comm.  When the model is
-                // created we don't know yet if the comm is valid so set_comm_live
-                // false.  Once we receive the first state push from the back-end
-                // we know the comm is alive.
-                return kernel.widget_manager.create_model({
-                    comm: new_comm,
-                    model_name: state[model_id].model_name,
-                    model_module: state[model_id].model_module,
-                }).then(function(model) {
-                    model.set_comm_live(false);
-                    return model._deserialize_state(state[model.id].state).then(function(state) {
-                        model.set_state(state);
-                        return model.request_state().then(function() {
-                            model.set_comm_live(true);
-                            return model;
+            var all_models = that._get_comm_list().then(function(live_comms) {
+                return Promise.all(_.map(Object.keys(state), function (model_id) {
+                    // Recreate a comm using the widget's model id. 
+                    var live = live_comms.hasOwnProperty(model_id);
+                    if (live) {
+                        var new_comm = new comm.Comm(kernel.widget_manager.comm_target_name, model_id);
+                        kernel.comm_manager.register_comm(new_comm);
+                    }
+                    return kernel.widget_manager.new_model({
+                        comm: new_comm,
+                        model_name: state[model_id].model_name,
+                        model_module: state[model_id].model_module,
+                        model_id: model_id,
+                    }).then(function(model) {
+                        return model._deserialize_state(state[model.id].state).then(function(state) {
+                            model.set_state(state);
+                            // Only request state for live comms
+                            if (live) {
+                                return model.request_state().then(function() {
+                                    return model;
+                                });
+                            } else {
+                               return model;
+                            }
```

Comment:
Regarding this section, one way to shorten would be to have `set_state` return the model, and `request_state` directly return the model if comm is dead, or return a promise for the model.

This would give 

`return model.set_state(state).request_state();`